File: testservice-web/src/services/notificationService.ts
import * as signalR from '@microsoft/signalr';

export interface Notification {
  type: 'schema_created' | 'schema_updated' | 'schema_deleted' | 'entity_created' | 'entity_updated' | 'entity_deleted';
  schemaName?: string;
  entityType?: string;
  entityId?: string;
  schema?: any;
  timestamp: string;
}

export type NotificationHandler = (notification: Notification) => void;

class NotificationService {
  private connection: signalR.HubConnection | null = null;
  private handlers: NotificationHandler[] = [];
  private reconnectTimer: NodeJS.Timeout | null = null;

  async connect() {
    const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:5000';
    const token = localStorage.getItem('token');
    
    console.log('?? Connecting to SignalR...');
    console.log('   API URL:', API_BASE_URL);
    console.log('   Has Token:', !!token);
    
    this.connection = new signalR.HubConnectionBuilder()
      .withUrl(`${API_BASE_URL}/notificationHub`, {
        skipNegotiation: true,
        transport: signalR.HttpTransportType.WebSockets,
        accessTokenFactory: () => {
          const token = localStorage.getItem('token');
          console.log('   Providing token to SignalR:', token ? 'Yes' : 'No');
          return token || '';
        }
      })
      .withAutomaticReconnect({
        nextRetryDelayInMilliseconds: (retryContext) => {
          // Exponential backoff: 0s, 2s, 10s, 30s...
          if (retryContext.previousRetryCount < 3) {
            return Math.pow(2, retryContext.previousRetryCount) * 1000;
          }
          return 30000; // Max 30 seconds
        }
      })
      .configureLogging(signalR.LogLevel.Information)
      .build();

    // Set up event handlers
    this.connection.on('SchemaCreated', (data: any) => {
      console.log('?? SchemaCreated event received from SignalR:', data);
      console.log('   Raw data:', JSON.stringify(data, null, 2));
      
      // Transform backend data to frontend Notification format
      const notification: Notification = {
        type: 'schema_created',
        schemaName: data.schemaName,
        timestamp: data.timestamp,
        schema: data.schema
      };
      
      console.log('   Transformed notification:', notification);
      this.notifyHandlers(notification);
    });

    this.connection.on('SchemaUpdated', (data: any) => {
      console.log('?? SchemaUpdated event received from SignalR:', data);
      console.log('   Raw data:', JSON.stringify(data, null, 2));
      
      // Transform backend data to frontend Notification format
      const notification: Notification = {
        type: 'schema_updated',
        schemaName: data.schemaName,
        timestamp: data.timestamp,
        schema: data.schema
      };
      
      console.log('   Transformed notification:', notification);
      this.notifyHandlers(notification);
    });

    this.connection.on('SchemaDeleted', (data: any) => {
      console.log('?? SchemaDeleted event received from SignalR:', data);
      console.log('   Raw data:', JSON.stringify(data, null, 2));
      
      // Transform backend data to frontend Notification format
      const notification: Notification = {
        type: 'schema_deleted',
        schemaName: data.schemaName,
        timestamp: data.timestamp
      };
      
      console.log('   Transformed notification:', notification);
      this.notifyHandlers(notification);
    });

    this.connection.on('EntityCreated', (data: Notification) => {
      console.log('Entity created notification:', data);
      this.notifyHandlers(data);
    });

    this.connection.on('EntityUpdated', (data: Notification) => {
      console.log('Entity updated notification:', data);
      this.notifyHandlers(data);
    });

    this.connection.on('EntityDeleted', (data: Notification) => {
      console.log('Entity deleted notification:', data);
      this.notifyHandlers(data);
    });

    this.connection.onreconnecting((error) => {
      console.warn('SignalR reconnecting...', error);
    });

    this.connection.onreconnected((connectionId) => {
      console.log('SignalR reconnected:', connectionId);
    });

    this.connection.onclose((error) => {
      console.error('SignalR connection closed:', error);
      // Try to reconnect after 5 seconds
      if (this.reconnectTimer) clearTimeout(this.reconnectTimer);
      this.reconnectTimer = setTimeout(() => {
        console.log('Attempting to reconnect...');
        this.connect().catch(console.error);
      }, 5000);
    });

    try {
      await this.connection.start();
      console.log('? SignalR connected successfully');
      console.log('   Connection ID:', this.connection.connectionId);
      console.log('   Connection State:', this.connection.state);
      
      // List all registered event handlers
      console.log('   Registered event handlers:', [
        'SchemaCreated',
        'SchemaUpdated', 
        'SchemaDeleted',
        'EntityCreated',
        'EntityUpdated',
        'EntityDeleted'
      ]);
    } catch (err) {
      console.error('? SignalR connection error:', err);
      // Retry after 5 seconds
      if (this.reconnectTimer) clearTimeout(this.reconnectTimer);
      this.reconnectTimer = setTimeout(() => {
        this.connect().catch(console.error);
      }, 5000);
    }
  }

  disconnect() {
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
    
    if (this.connection) {
      this.connection.stop();
      this.connection = null;
    }
  }

  subscribe(handler: NotificationHandler): () => void {
    this.handlers.push(handler);
    
    // Return unsubscribe function
    return () => {
      this.handlers = this.handlers.filter(h => h !== handler);
    };
  }

  private notifyHandlers(notification: Notification) {
    this.handlers.forEach(handler => {
      try {
        handler(notification);
      } catch (err) {
        console.error('Error in notification handler:', err);
      }
    });
  }

  getConnectionState(): signalR.HubConnectionState | null {
    return this.connection?.state ?? null;
  }

  isConnected(): boolean {
    return this.connection?.state === signalR.HubConnectionState.Connected;
  }

  // Debug method to check connection status
  getDebugInfo() {
    return {
      connectionState: this.connection?.state,
      connectionId: this.connection?.connectionId,
      isConnected: this.isConnected(),
      handlerCount: this.handlers.length
    };
  }
}

export const notificationService = new NotificationService();
export default notificationService;

File: TestService.Api/Services/NotificationService.cs
using Microsoft.AspNetCore.SignalR;
using TestService.Api.Hubs;

namespace TestService.Api.Services;

public class NotificationService : INotificationService
{
    private readonly IHubContext<NotificationHub> _hubContext;
    private readonly ILogger<NotificationService> _logger;

    public NotificationService(
        IHubContext<NotificationHub> hubContext,
        ILogger<NotificationService> logger)
    {
        _hubContext = hubContext;
        _logger = logger;
    }

    public async Task NotifySchemaCreated(string schemaName, object schema)
    {
        try
        {
            await _hubContext.Clients.All.SendAsync("SchemaCreated", new
            {
                type = "schema_created",
                schemaName,
                schema,
                timestamp = DateTime.UtcNow
            });

            _logger.LogInformation("Notification sent: Schema created - {SchemaName}", schemaName);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to send schema created notification for {SchemaName}", schemaName);
        }
    }

    public async Task NotifySchemaUpdated(string schemaName, object schema)
    {
        try
        {
            await _hubContext.Clients.All.SendAsync("SchemaUpdated", new
            {
                type = "schema_updated",
                schemaName,
                schema,
                timestamp = DateTime.UtcNow
            });

            _logger.LogInformation("Notification sent: Schema updated - {SchemaName}", schemaName);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to send schema updated notification for {SchemaName}", schemaName);
        }
    }

    public async Task NotifySchemaDeleted(string schemaName)
    {
        try
        {
            await _hubContext.Clients.All.SendAsync("SchemaDeleted", new
            {
                type = "schema_deleted",
                schemaName,
                timestamp = DateTime.UtcNow
            });

            _logger.LogInformation("Notification sent: Schema deleted - {SchemaName}", schemaName);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to send schema deleted notification for {SchemaName}", schemaName);
        }
    }

    public async Task NotifyEntityCreated(string entityType, string entityId)
    {
        try
        {
            await _hubContext.Clients.All.SendAsync("EntityCreated", new
            {
                type = "entity_created",
                entityType,
                entityId,
                timestamp = DateTime.UtcNow
            });

            _logger.LogInformation("Notification sent: Entity created - {EntityType}/{EntityId}", entityType, entityId);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to send entity created notification for {EntityType}/{EntityId}", entityType, entityId);
        }
    }

    public async Task NotifyEntityUpdated(string entityType, string entityId)
    {
        try
        {
            await _hubContext.Clients.All.SendAsync("EntityUpdated", new
            {
                type = "entity_updated",
                entityType,
                entityId,
                timestamp = DateTime.UtcNow
            });

            _logger.LogInformation("Notification sent: Entity updated - {EntityType}/{EntityId}", entityType, entityId);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to send entity updated notification for {EntityType}/{EntityId}", entityType, entityId);
        }
    }

    public async Task NotifyEntityDeleted(string entityType, string entityId)
    {
        try
        {
            await _hubContext.Clients.All.SendAsync("EntityDeleted", new
            {
                type = "entity_deleted",
                entityType,
                entityId,
                timestamp = DateTime.UtcNow
            });

            _logger.LogInformation("Notification sent: Entity deleted - {EntityType}/{EntityId}", entityType, entityId);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to send entity deleted notification for {EntityType}/{EntityId}", entityType, entityId);
        }
    }
}
